Extend form for creating new cloud volumes

gberginc opened the pull request:
Current form for creating a new cloud volume is very specific to OpenStack and allows specification of volume name, size and target tenant. With this change we are planning on adding dynamic support for provisioning of new cloud volumes allowing for different backend providers (OpenStack Cinder and Amazon EBS for the time being).

Current video of the UI: http://x.k00.fr/lqzwd

Current appearance for Amazon:
![screen shot 2017-02-28 at 22 49 57](https://cloud.githubusercontent.com/assets/1437960/23445990/ebcd0cc8-fe3f-11e6-9340-d45bafbd8d61.png)

Current appearance for OpenStack:
![screen shot 2017-02-28 at 22 50 03](https://cloud.githubusercontent.com/assets/1437960/23446002/f7a296da-fe3f-11e6-8f85-e5fce0d66015.png)


@miq-bot add_label wip,storage,question

@AparnaKarve, @dclarizio This is the first attempt to support the Amazon block storage provider along side OpenStack ([ref doc](https://docs.google.com/document/d/1Idt4xqMEmm14jUjQ_IqIgGjVO6oG4YZcKzeQM8y5wcI/edit)). Could you please have a look if this is going in the right direction so that we can add other fields for Amazon provider?

/cc @miha-plesko

gberginc commented on app/assets/javascripts/controllers/cloud_volume/cloud_volume_form_controller.js:
```diff
@@ -26,6 +31,25 @@ ManageIQ.angular.app.controller('cloudVolumeFormController', ['$http', '$scope',
+  $scope.storageManagerChanged = function (id) {
+    miqService.sparkleOn();
+    
+    $http.get('/cloud_volume/storage_manager_form_fields/' + id)
```
I've seen [here](https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/assets/javascripts/controllers/cloud_subnet/cloud_subnet_form_controller.js#L65-L71) that multiple requests are made to the backend, but decided to keep it simple for now.

Is it mandatory to request all field values separately or is it ok to have one function that will provide only the set of those choices that are applicable to the actual provider type?

AparnaKarve replied:
A single call is preferred since that would mean a single trip to the server (preferred) as opposed to multiple requests (not preferred).

Going forward, we want all our GET and POST requests to be actual API calls, which means no new routes have to be added or no new UI controller code to support that route.

gberginc replied:
Should I look for this now or is it planned for future releases? That would complicate things a bit and I think definitely require multiple calls. Is there perhaps an example of this approach already somewhere in the code?

gberginc commented on app/assets/javascripts/controllers/cloud_volume/cloud_volume_form_controller.js:
```diff
@@ -26,6 +31,25 @@ ManageIQ.angular.app.controller('cloudVolumeFormController', ['$http', '$scope',
+    var data = response.data;
+
+    $scope.afterGet = true;
+    $scope.cloudVolumeModel.emstype = data.emstype;
```
Honestly, I was having a bit of a problem what should be put inside `cloudVolumeModel` and what can be put directly in the `$scope`. I guess it doesn't really matter, but if you have preference, please let me know.

AparnaKarve replied:
`cloudVolumeModel` should constitute all the form data.

`$scope.cloudVolumeModel.emstype = data.emstype;` looks correct.

gberginc replied:
Great. What about the next two lines where we just set the choices for the UI. I don't consider them being part of the model, but rather just supporting the UI, so I decided not to add them into model.

gberginc commented on app/controllers/cloud_volume_controller.rb:
```diff
@@ -129,6 +129,22 @@ def cloud_volume_form_fields
+    form_fields = { :emstype => ems.emstype }
+
+    case ems.type
+    when "ManageIQ::Providers::Amazon::StorageManager::Ebs"
```
Is this ok to do a check like this in the UI? Please note that I expect further form fields for Amazon, but not for Openstack (for now, because also the existing code only allows cloud tenant).

gberginc commented on app/controllers/cloud_volume_controller.rb:
```diff
@@ -129,6 +129,22 @@ def cloud_volume_form_fields
+    assert_privileges("cloud_volume_new")
+    ems = find_by_id_filtered(ExtManagementSystem, params[:id])
+
+    form_fields = { :emstype => ems.emstype }
```
So, I need to know what type the provider is. I used the approach of the `ems_common_form_controller` where the `emstype` was set.

gberginc commented on app/controllers/cloud_volume_controller.rb:
```diff
@@ -239,8 +255,8 @@ def new
     @in_a_form = true
-    @cloud_tenant_choices = {}
-    CloudTenant.all.each { |tenant| @cloud_tenant_choices[tenant.name] = tenant.id }
+    @storage_manager_choices = {}
```
Here we are returning just the first level choices - only after we choose the storage manager, we can get the rest. I am not actually sure how this works for OpenStack, because it will return all cloud tenants, regardless of the cloud manager (i.e. if I have several OpenStack managers, I'll get these cloud tenants mixed).

AparnaKarve commented:
@gberginc I just did a cursory review and this looks great for a first pass!
(I will do a detailed review shortly)

Couple things: 
- Could you attach a UI screenshot here?
- As our next step, we should look into replacing this call `$http.get('/cloud_volume/storage_manager_form_fields/' + id)` with an actual API call.

gberginc commented:
@AparnaKarve I've added two screenshots to the PR (I had them yesterday already but forgot to attach them).

Regarding the API change I'd like to check something first. Current backend handler gets the EMS ID of the chosen storage manager. Based on the type it decides what needs to be collected and returned to the frontend. If we replace all this with API, I think we should do the following:

1. Set the `emstype` in the initial handler for the `new` (currently, we only get `id` and `name`).
2. Once the manager is changed, based on a JS (angular controller) `emstype` we decide for which fields we need additional choices
3. For all these choices we do a separate API call. Since API call provides complete objects, we then need to filter the required data in angular response handler

Is there a different way?

gberginc commented:
@AparnaKarve I've added a second commit with more fields for AWS. Here's the video showing how the form works at the moment: http://x.k00.fr/lqzwd.

For starters that's it. I should add some validation rules (for example, there are min/max for different fields).

Commits are going to be tidied in the end (I separated it so that you'll be able to see the new changes easier).

sergio-ocon commented:
@AparnaKarve would it be good to merge it as is, and then open a second use case to add the pending elements? We could update pivotal with it and we would have the functionality working

AparnaKarve commented:
@sergio-ocon, I'm currently testing this in an environment that @gberginc gave me to test on...
And from what I'm seeing so far the form isn't quite functional yet, and very much W-I-P.

The dropdowns work fine, yes.
However, we need to use REST API behind the scenes to populate them.

sergio-ocon commented:
Thanks @AparnaKarve. I thought it was functional, although not complete.

gberginc commented:
@AparnaKarve I've now added support for getting required form fields for openstack and amazon and tested this with the actual provider.

Video showing how the form works for OpenStack and Amazon: http://x.k00.fr/vsld4. There is one interesting bug (?) for OpenStack: when I create a volume, it's added with status `creating`. Once the volume is created, the parser will create a new object in the database with correct status. Probably something for core team @blomquisg? This doesn't happen for Amazon (as seen at the end of the video).

AparnaKarve commented on app/controllers/cloud_volume_controller.rb:
```diff
@@ -129,6 +129,22 @@ def cloud_volume_form_fields
+      form_fields[:availability_zone_choices] = ems.availability_zones.each_with_object([]) { |az, choices| choices << az.name }.sort
+    when "ManageIQ::Providers::StorageManager::CinderManager"
+      form_fields[:cloud_tenant_choices] = ems.parent_manager.cloud_tenants.each_with_object({}) { |tenant, choices| choices[tenant.name] = tenant.id }
+    end
```
@gberginc I'm looking for a possible REST API to replace the above.

With an API like that:
```
/api/providers/<id>?attributes=type,parent_manager.availability_zones,parent_manager.cloud_tenants
```
you can retrieve all 3 form fields in one go.

But just wondering -- why is it that for `availability_zones` you don't need a `parent_manager` and for `cloud_tenants` you do?

gberginc replied:
This seems like it should work!

There is a difference in the way Amazon and OpenStack managers work. First, [here](https://github.com/ManageIQ/manageiq-ui-classic/pull/517/files#diff-6e783b8caed5140ba7bdfad51acb5f6cR259) we are retrieving managers that have `supports_block_storage` set.

For Amazon, this is set in [EBS block storage manager](https://github.com/ManageIQ/manageiq-providers-amazon/blob/master/app/models/manageiq/providers/amazon/storage_manager/ebs.rb#L9). For OpenStack in the [Cinder manager](https://github.com/ManageIQ/manageiq/blob/master/app/models/manageiq/providers/storage_manager/cinder_manager.rb#L11).

EBS delegates retrieval of availability zones to the parent manager because it knows its parent manager is the Amazon cloud manager. On the other hand, Cinder manager is more generic because it can be used by multiple different cloud managers (not currently, but in theory). Therefore, there was no delegation to the parent made so far (and I didn't want to add it just for the sake of using it here).

That being said, as you've seen while preparing the API call, explicitly including the `parent_manager` in case of EBS also works so I think we can use them in both cases to not cause confusion.

AparnaKarve replied:
Excellent!! I'm glad that the suggested API works!

gberginc commented on app/assets/javascripts/controllers/cloud_volume/cloud_volume_form_controller.js:
```diff
@@ -1,10 +1,26 @@
+    { type: "sc1", name: "Cold HDD (SC1)" },
+    { type: "magnetic", name: "Magnetic" }
+  ];
+
```
Yes, it's fixed. The question is whether this is something the provider should offer. Something like the list of regions [here](https://github.com/ManageIQ/manageiq-providers-amazon/blob/master/app/models/manageiq/providers/amazon/regions.rb).

gberginc commented:
@AparnaKarve The latest commit updates the code to use existing APIs.

gberginc commented:
[This commit](b1b1b2409a5b104e84c06275fd14fb6631db243b) also fixes your other comments. I plan to squash these four commits after we are fine with the changes.

Two more things:
* Do you think I could now remove the WIP?
* What do you think I should add tests for? I didn't find a spec test for the "create volume". Should we add tests for angular as well? If so, I'd appreciate any pointer on how to approach this.

AparnaKarve commented:
> Do you think I could now remove the WIP?

@gberginc Like you said, tests! 
Yes, we need some tests before we can un-WIP.

Also, let me play with this a little more today, just to check if everything in the form is in order.

gberginc commented:
@AparnaKarve In the last commit, I've added the tests for the controller. I've also changed the `create_volume` to `create_volume_queue` because I don't think this operation should be synchronous. I've tested with OpenStack and Amazon.

AparnaKarve commented:
> I've also changed the create_volume to create_volume_queue because I don't think this operation should be synchronous

Nice!

Will review the tests today.

gberginc commented:
@AparnaKarve [this one](https://github.com/ManageIQ/manageiq-ui-classic/pull/600) might also be of interest to you. It resolves the problem of requiring OpenStack provider for testing this PR because the button has been disabled if only Amazon provider was available.

AparnaKarve commented:
@gberginc Rspec tests look good.

-----------
Noticed a couple things while testing the form:
- The Edit form displays only the Name field, not the other fields
<img width="1436" alt="screen shot 2017-03-06 at 5 06 01 pm" src="https://cloud.githubusercontent.com/assets/1538216/23637207/3a5a2cce-028f-11e7-853d-602b1648dd1c.png">

------------

As far as the `New` form goes, there are a few things to address purely from a UX standpoint
- The `Placement` label and the UI fields underneath it should be visible only upon selecting the `Storage Manager` from the dropdown.
- `<Choose>` should be disabled once an item is chosen from the dropdown
- Required fields should be clearly marked as Required
- Use `bs-switch` for the Encryption checkbox

-------------
One other thing I noticed was -- 
in the AWS console, the volume is showing as `available`
<img width="1235" alt="screen shot 2017-03-06 at 5 08 24 pm" src="https://cloud.githubusercontent.com/assets/1538216/23637273/a4867b48-028f-11e7-8cf5-df6e2f25c927.png">

However, in the `Cloud Volume` list view, it shows as `creating` even after several minutes had elapsed after creating the volume.
<img width="916" alt="screen shot 2017-03-06 at 5 10 24 pm" src="https://cloud.githubusercontent.com/assets/1538216/23637335/f10d4820-028f-11e7-8348-85f469a9fcfa.png">

gberginc commented:
Thanks @AparnaKarve for this review! Below are some comments.

> The Edit form displays only the Name field, not the other fields

I was hoping to make this part of a separate PR, in particular because OpenStack will still allow changing only the name.
 
> As far as the New form goes, there are a few things to address purely from a UX standpoint
> 
> The Placement label and the UI fields underneath it should be visible only upon selecting the Storage Manager from the dropdown.

Done with the last commit. Nevertheless, I still see the "flickering" which I am not sure if it's just my problem or something actually is wrong with the way form is structured/handled by Angular.

> <Choose> should be disabled once an item is chosen from the dropdown

Done.

> Required fields should be clearly marked as Required

This one is a bit problematic to me. Can you please elaborate a bit more, because I am not sure if you wanted to display `Required` for all the required fields or something else?

I've done plenty of

```
        %span.help-block{"ng-show" => "angularForm.storage_manager_id.$error.required"}
          = _("Required")
```

but I don't like how it behaves. For example, when the form starts, all selects have this displayed, but for example the `name` does not. What I don't like is that I had to remove model initialization: https://github.com/ManageIQ/manageiq-ui-classic/pull/517/commits/92961b57c298f42c0bd34fe146d0efcceba4be9e#diff-638665f4b337f59cc291a102efee0ce6L3 (but I am not sure if it was done right before)

> Use bs-switch for the Encryption checkbox

Done.

> One other thing I noticed was --
> in the AWS console, the volume is showing as available
> However, in the Cloud Volume list view, it shows as creating even after several minutes have elapsed after creating the volume.

The reason for this is that the worker did not refresh the cloud volume states. Navigate to your Compute > Clouds > Providers and refresh it manually. I haven't tried this yet, but [this doc](https://github.com/ManageIQ/guides/blob/master/providers/amazon_aws_config.md) describes how to configure AWS Config to get the events.

gberginc commented:
Slowly getting there :smile:. Hopefully, we'll soon remove the WIP and squash these commits.

Commit 861f0e1 now provides the information for the required fields. Video with all the changes (including those from the previous commits): http://x.k00.fr/jvikg.

Regarding the flickering problem, please notice in the video between 3rd and 4th second. Here is the frame:

<img width="594" alt="screen shot 2017-03-07 at 06 50 16" src="https://cloud.githubusercontent.com/assets/1437960/23643524/9e460adc-0302-11e7-8eec-1851913310e0.png">

Please notice how the fields are the ones used in the browser by default. Thus, I think something kicks in too late, but I just don't know what.

AparnaKarve commented:
@gberginc Thanks for that video -- it explains things so much better!

The Add form, btw, is looking so much presentable now. Thanks for looking into all the UX issues I mentioned above.

I will take a shot at the flickering issue today.

AparnaKarve commented:
@gberginc I have good news!
The flickering issue can be resolved by adding `"ng-cloak" => ""` to the form. So just add it here --
```haml
%form#form_div{:name => "angularForm", 'ng-controller' => "cloudVolumeFormController", "ng-cloak" => ""}
```

AparnaKarve commented:
Also, a few more improvements, some of which are in fact just more than improvements since they actually address a bug in the form.

- Remove all instances of `miqrequired`. This was a directive written a while back and we don't use it anymore.
`required` alone should suffice.
- Either remove the `'ng-maxlength' => 10` requirement for `aws_iops` or increase it to 50 or something. The form would become invalid when `Not Applicable` is displayed in that field, and `Add` button would never light up.
(`Not Applicable` characters length exceeds 10)
- Can we improve the overall flow of the form? As in, make it flow more logically.
For instance, would it be possible to move `Size` right under `IOPS`, since changing `Size` directly affects what we display in `IOPS`?

gberginc commented:
Thanks for the `ng-cloak` thing. It's interesting that this is only used for cloud volume forms - any special reason?

Regarding other comments, see below:

> Remove all instances of miqrequired. This was a directive written a while back and we don't use it anymore.
> required alone should suffice.

Removed in the last commit.

> Either remove the 'ng-maxlength' => 10 requirement for aws_iops or increase it to 50 or something. The form would become invalid when Not Applicable is displayed in that field, and Add button would never light up.
> (Not Applicable character length exceeds 10)

Thanks for spotting this! Fixed in the commit.

> Can we improve the overall flow of the form? As in, make it flow more logically.
> For instance, would it be possible to move Size right under IOPS, since changing Size directly affects what we display in IOPS?

You are correct about this. Now we have two groups: Basic info and placement. Since volume type, IOPS, encryption are not actually "Placement" information I was actually thinking about removing the Placement group altogether. The flow would then be as follows

**For OpenStack provider**
 * Storage Manager (OpenStack)
 * Cloud Tenant
 * Volume Name
 * Size

**For Amazon provider**
 * Storage Manager (OpenStack)
 * Availability zone
 * Volume Name
 * Volume type
 * Size
 * IOPS
 * Encryption

The fields would be flat, but would appear depending on the chosen provider.

After this, I'd like to squash the commits to clean the history a bit.

gberginc commented:
@AparnaKarve I've rebased this PR and squashed the commits. After reordering, the form looks like:

![screen shot 2017-03-08 at 06 01 08](https://cloud.githubusercontent.com/assets/1437960/23690805/bb773bc2-03c4-11e7-89d8-88d34550bfa9.png)
![screen shot 2017-03-08 at 06 01 13](https://cloud.githubusercontent.com/assets/1437960/23690806/bb9dfe7e-03c4-11e7-8cd2-8f404b9ed4f5.png)

Note that I've removed `Basic Information` heading at the top because we now only have one group of information.

@miq-bot remove_label wip

gberginc commented:
I've extended the spec to include the encryption on EBS volumes as well (while cleaning up for the final push, I noticed I missed the encryption propagation in the controller).

gberginc commented:
@miq-bot remove_label question

gberginc commented:
@AparnaKarve @dclarizio I have pushed one more commit resolving ` Extra empty line detected at block body end.` message from rubocop. I think other warnings are usually ignored (in particular those causing codeclimate to fail).

Is there anything else you'd like to see changed within this PR?

AparnaKarve commented:
@gberginc It looks like all the issues that we discussed here in this PR and in gitter have been resolved, which is great.
Let me know if this is correct -- things that we plan to address in a follow-up PR are : 
- Edit Cloud Volumes
- Using `API.post` to save the Cloud Volume records

gberginc commented:
@AparnaKarve This is correct. I also plan to finish adding support for creating cloud volumes to other views (for example, the view `ems_storage` that shows provider's cloud volumes).

AparnaKarve commented:
@dclarizio I've tested this in the UI for Adding new Cloud Volumes and it works very well.

As indicated above https://github.com/ManageIQ/manageiq-ui-classic/pull/517#issuecomment-285540361, https://github.com/ManageIQ/manageiq-ui-classic/pull/517#issuecomment-285594492, there are a couple things that will be addressed in follow-up PR(s)

This is a good first step for the Cloud Volume form, and I think we should merge the code at this point to get the basic functionality in. Thanks.

sergio-ocon commented:
👍
